### ISSUE TITLE
Support Mistral-Nemo-Instruct-2407 128K

### ISSUE THREAD

**Issue by mirek190** [ggerganov/llama.cpp#8577]
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Any plans for support Mistral-Nemo-Instruct-2407 128K ?

https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407

### Motivation

enhancement

**Comment by 0wwafa** [comment]
yes, please. this one is going to be good and soon finetunes will start to popup...

**Comment by 0wwafa** [comment]
And this: https://huggingface.co/nvidia/Mistral-NeMo-12B-Instruct

**Comment by delphijb** [comment]
I second the request. This model is likely to become the benchmark for the 7-12b segment, and finetuning will indeed appear rapidly. Thx in advance

**Comment by stduhpf** [comment]
They claim it can be a drop-in replacement of Mistral 7B, so surely it shouldn't be too much work to make it work with ggml since Mistral 7B works.

**Comment by EliEron** [comment]
> They claim it can be a drop-in replacement of Mistral 7B, so surely it shouldn't be too much work to make it work with ggml since Mistral 7B works.

The issue is that it uses a custom tokenizer named Tekken. That's not an issue for any program that uses the transformers tokenizer system since it supports the custom tokenizer. Which is why they call it a drop in replacement.

Llama.cpp however does not use that system, custom tokenizers have to be implemented manually. And implementing new tokenizers correctly is usually not easy. Gemma and Llama-3's tokenizer for instance took quite a while to implement properly, and it took multiple attempts to do so as bugs where found over time.

**Comment by iamlemec** [comment]
I actually think the tokenizer might not be too different from others. It's listed as `GPT2Tokenizer` in the `tokenizer_config.json` and it has a pre-tokenizer of the usual form. I was able to add it in the standard fashion with pre-tokenizer and the update script.

The other issue is that the tensors shapes relating to attention are not the sizes expected by the current implementation of Mistral (see my other comment here https://github.com/ggerganov/llama.cpp/discussions/8576#discussioncomment-10088864). I was able to brute force hack it into at least running, and I'm getting sensible output, which makes me think the tokenizer is doing okay. For example:

```
PROMPT: What is the capital of South Korea in Hangul?
RESPONSE: The capital of South Korea in Hangul is 서울 (Seoul).
```

**Comment by netrunnereve** [comment]
If this model works well we should also try to add FP8 support into llama.cpp and make full use of the QAT. That will take more work to compute compared to Q8_0 without native FP8 support but it'll probably end up being memory bound anyways.

**Comment by iamlemec** [comment]
For those curious, I have a seemingly working implementation here: https://github.com/iamlemec/llama.cpp/tree/mistral-nemo. Will push a PR for it once #8579 is merged and I can rebase off of it.

**Comment by mirek190** [comment]
> For those curious, I have a seemingly working (and now non-hacky) inference implementation here: https://github.com/iamlemec/llama.cpp/tree/mistral-nemo. Will push a PR for it once #8579 is merged and I can rebase off of it.

Can't wait to test ;)

**Comment by foldl** [comment]
Also, for those who are interested, [chatllm.cpp](https://github.com/foldl/chatllm.cpp) supports this.

**Comment by stduhpf** [comment]
> For those curious, I have a seemingly working (and now non-hacky) inference implementation here: https://github.com/iamlemec/llama.cpp/tree/mistral-nemo. Will push a PR for it once #8579 is merged and I can rebase off of it.

Seems to work perfectly so far. Nice job.

**Comment by muhammadyusuf-kurbonov** [comment]
#8579 is merged

**Comment by legraphista** [comment]
Just quantized `Mistral-Nemo-Instruct` and trying to run it I get the following error:
```
llm_load_tensors: ggml ctx size =    0.17 MiB
llama_model_load: error loading model: check_tensor_dims: tensor 'blk.0.attn_q.weight' has wrong shape; expected  5120,  5120, got  5120,  4096,     1,     1
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model './Mistral-Nemo-Instruct-2407.Q8_0.gguf'
main: error: unable to load model
```

Looks like there's a shape mismatch. 

According to the config file, the hidden size should be 5120 https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407/blob/main/config.json#L10

**Comment by mirek190** [comment]
where I can find a proper gguf?

**Comment by maziyarpanahi** [comment]
Hi @legraphista 

I have a new build from the main branch with the new PR merged, I am also using `convert_hf_to_gguf.py`. But I am getting this error:

```
llama.cpp/gguf-py/gguf/metadata.py", line 233, in get_model_id_components
    if at_start and ((len(t) == 0 and part[0].isalpha()) or "version" in t):
IndexError: string index out of range
```

Could you please let me know if I am missing something here?

**Comment by stduhpf** [comment]
> Hi @legraphista
> 
> I have a new build from the main branch with the new PR merged, I am also using `convert_hf_to_gguf.py`. But I am getting this error:
> 
> ```
> llama.cpp/gguf-py/gguf/metadata.py", line 233, in get_model_id_components
>     if at_start and ((len(t) == 0 and part[0].isalpha()) or "version" in t):
> IndexError: string index out of range
> ```
> 
> Could you please let me know if I am missing something here?

Only the tokenizer support has been added by the PR. If you want to run the model, you çan use @iamlemec's fork.

**Comment by iamlemec** [comment]
@mirek190 I have a `Q5_K` of Instruct here: https://huggingface.co/CompendiumLabs/mistral-nemo-instruct-2407-gguf. Happy to add more variants if needed.

**Comment by maziyarpanahi** [comment]
> > Hi @legraphista
> > I have a new build from the main branch with the new PR merged, I am also using `convert_hf_to_gguf.py`. But I am getting this error:
> > ```
> > llama.cpp/gguf-py/gguf/metadata.py", line 233, in get_model_id_components
> >     if at_start and ((len(t) == 0 and part[0].isalpha()) or "version" in t):
> > IndexError: string index out of range
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you please let me know if I am missing something here?
> 
> Only the tokenizer support has been added by the PR. If you want to run the model, you çan use @iamlemec's fork.

I am actually trying to quantize it at the moment, since I saw it happened successfully here, so I was wondering.

**Comment by mirek190** [comment]
> @mirek190 I have a `Q5_K` of Instruct here: https://huggingface.co/CompendiumLabs/mistral-nemo-instruct-2407-gguf. Happy to add more variants if needed.

````
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3090, compute capability 8.6, VMM: yes
llm_load_tensors: ggml ctx size =    0.34 MiB
llama_model_load: error loading model: check_tensor_dims: tensor 'blk.0.attn_q.weight' has wrong shape; expected  5120,  5120, got  5120,  4096,     1,     1
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model 'models/new3/mistral-nemo-instruct-q5_k.gguf'
main: error: unable to load model
````

You version also doesn't work

**Comment by cwillu** [comment]
@mirek190 try running a make clean first; the project makefiles don't appear to be 100% reliable.

**Comment by ubergarm** [comment]
I got @iamlemec 's commit working with his specific GGUF like so [CompendiumLabs/mistral-nemo-instruct-2407-gguf](https://huggingface.co/CompendiumLabs/mistral-nemo-instruct-2407-gguf/discussions/1#669c2c7f6814a4af546b3b71)

Didn't get it working with [QuantFactory/Mistral-Nemo-Instruct-2407-GGUF](https://huggingface.co/QuantFactory/Mistral-Nemo-Instruct-2407-GGUF/discussions/1#669c282fc9111326dcfc36fb)

Looks like PR is incoming #8604

**Comment by mlsterpr0** [comment]
> Also, for those who are interested, [chatllm.cpp](https://github.com/foldl/chatllm.cpp) supports this.

Yeah, cool, but how do you run it, if the name of the model is nowhere to find? python chatllm.py -i -m :?????????

**Comment by muhammadyusuf-kurbonov** [comment]
#8604 is merged!!

**Comment by danilofalcao** [comment]
```
llama_model_load: error loading model: error loading model vocabulary: unknown pre-tokenizer type: 'tekken'
llama_load_model_from_file: failed to load model
Traceback (most recent call last):
  File "/root/Mistral-Nemo-Instruct-2407/chat.py", line 21, in <module>
    generated_text = run_inference(model_path, prompt)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/Mistral-Nemo-Instruct-2407/chat.py", line 5, in run_inference
    llm = Llama(model_path=model_path)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/env/lib/python3.11/site-packages/llama_cpp/llama.py", line 358, in __init__
    self._model = self._stack.enter_context(contextlib.closing(_LlamaModel(
                                                               ^^^^^^^^^^^^
  File "/root/env/lib/python3.11/site-packages/llama_cpp/_internals.py", line 54, in __init__
    raise ValueError(f"Failed to load model from file: {path_model}")
ValueError: Failed to load model from file: ggml-model-Q4_K_M.gguf
```

it's able to convert but still having issues with pre-tokenizer tekken using llama-cpp-python

**Comment by aashish-1904** [comment]
Latest quants are available at https://huggingface.co/QuantFactory/Mistral-Nemo-Instruct-2407-GGUF

> I got @iamlemec 's commit working with his specific GGUF like so [CompendiumLabs/mistral-nemo-instruct-2407-gguf](https://huggingface.co/CompendiumLabs/mistral-nemo-instruct-2407-gguf/discussions/1#669c2c7f6814a4af546b3b71)
> 
> Didn't get it working with [QuantFactory/Mistral-Nemo-Instruct-2407-GGUF](https://huggingface.co/QuantFactory/Mistral-Nemo-Instruct-2407-GGUF/discussions/1#669c282fc9111326dcfc36fb)
> 
> Looks like PR is incoming #8604

Updating the quants based on the latest release at: [Mistral-Nemo-Instruct-2407-GGUF](https://huggingface.co/QuantFactory/Mistral-Nemo-Instruct-2407-GGUF)

**Comment by EricGrange** [comment]
Just tried with latest quants at [QuantFactory/Mistral-Nemo-Instruct-2407-GGUF](https://huggingface.co/QuantFactory/Mistral-Nemo-Instruct-2407-GGUF/discussions/1#669c282fc9111326dcfc36fb) and build 3437, getting the following error (Xeon E-2176G with 64GB RAM, under Debian). Apparently it's trying to allocate a very large CPU buffer

```
ggml_backend_cpu_buffer_type_alloc_buffer: failed to allocate buffer of size 167772160032
llama_kv_cache_init: failed to allocate buffer for kv cache
llama_new_context_with_model: llama_kv_cache_init() failed for self-attention cache
llama_init_from_gpt_params: error: failed to create context with model 'Mistral-Nemo-Instruct-2407.Q4_K_S.gguf'
 ERR [              load_model] unable to load model | tid="140094037281280" timestamp=1721650251 model="Mistral-Nemo-Instruct-2407.Q4_K_S.gguf"

```

**Comment by sbelenki** [comment]
> Just tried with latest quants at [QuantFactory/Mistral-Nemo-Instruct-2407-GGUF](https://huggingface.co/QuantFactory/Mistral-Nemo-Instruct-2407-GGUF/discussions/1#669c282fc9111326dcfc36fb) and build 3437, getting the following error (Xeon E-2176G with 64GB RAM, under Debian). Apparently it's trying to allocate a very large CPU buffer
> 
> ```

I have encountered the same problem while testing new build, [the advice that helped me](https://github.com/ggerganov/llama.cpp/pull/8604#issuecomment-2241367454) was to use `-c` parameter

**Comment by EricGrange** [comment]
Thanks confirmed, passing the context size explicitly does the trick.
It seems to work correctly also when using "-c 131072" (128k if I'm not mistaken)

**Comment by jthack** [comment]
i did an ollama update and still getting the same error.

**Comment by MoonRide303** [comment]
Initial tests with small context (`-c 8192`) look good to me - no issues observed so far. Both conversion and inference with llama.cpp b3438.

**Comment by 0wwafa** [comment]
tested. seems to work.

**Comment by mirek190** [comment]
How long text? 
I tested on 8k context text and failed badly.

**Comment by mirek190** [comment]
still testing .. I think -fa ( flash attention makes problems - degradation and performance loss ) as this model has huge ctx I used -fa

**Comment by foldl** [comment]
I think this model uses SWA. The memory requirement is much less if SWA is used.

I have tested it with 4.7k tokens (using SWA), and it looks ok.

**Comment by mirek190** [comment]
> I think this model uses SWA. The memory requirement is much less if SWA is used.
> 
> I have tested it with 4.7k tokens (using SWA), and it looks ok.

4.7 k tokens with 128k model ... yes very productive....

**Comment by foldl** [comment]
@mirek190 the point is that with SWA, 128k context length won't blow up your memory.

**Comment by ehartford** [comment]
I already saw this line in convert_hf_to_gguf_update.py

    {"name": "tekken",         "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/mistralai/Mistral-Nemo-Base-2407", },
  
I already executed convert_hf_to_gguf_update.py

but still it doesn't work.

As a guess, I tried this:

python ./convert_hf_to_gguf.py /workspace/axolotl/dolphin-2.9.3-mistral-nemo-hf --model-name tekken

But that didnt' work either

**Comment by mirek190** [comment]
Yesterday I used Bartowski gguf llama 3.1 q8   

test perplexity : no difference if I set ctx 8192 or 100000 same result 

````
build\bin\llama-perplexity.exe --model models/new3/Meta-Llama-3.1-8B-Instruct-imatrix.q8_0.gguf --threads 30 -ngl 99 --hellaswag --hellaswag-tasks 400 -f models\hellaswag_val_full.txt -c 8192
````
result - 79.25

Today used gguf imatrix from here  https://huggingface.co/AI-Engine/Meta-Llama-3.1-8B-Instruct-GGUF/tree/main
clams "Use the -imatrix versions (they use [imatrix](https://github.com/ggerganov/llama.cpp/discussions/5263#discussioncomment-8395384) and the bpe-llama tokenizer which should theoretically improve the output)"

result - 80.50

But still mostly wrong answers something 3/10 times is correct  

````
I have 10 apples. I find 3 gold coins in the bottom of a river. The river runs near a big city that has something to do with what I can spend the coins on. I then lose 4 apples but gain a gold coin. Three birds run into my path and drop 6 apples each. I play an online game and win 6 gold coins but I have to share them equally with my 2 teammates. I buy apples for all the coins I have. The price of an apple is 0.5 coins. How many apples do I have? And where is the river?
````
answer 36
or

````
If my BMI is 20.5 and my height is 172cm, how much would I weigh if I gained 5% of my current weight? 
````
answer 63.68kg   -

Lllama 8.1 8b with https://groq.com/ is giving here something between 63.4 - 63.5 what is excellent result local llm (build [b3452](https://github.com/ggerganov/llama.cpp/releases/tag/b3452))  WHEN is "almost" correct have something 62.3 or 64.4 but mostly 0.6 kg or 137.2 kg ...

But https://groq.com/   with llama 3.1 8b always answers those questions correctly 10/10  times

**Comment by iamlemec** [comment]
@ehartford They made some changes to `tokenizer_config.json` a day or two after release. See the commit here: https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407/commit/dac9c9e98f83322b32e32b48c118f079930772d6. Updating yours similarly should make the checksum match up.

**Comment by ehartford** [comment]
Thanks I'll do that

**Comment by rkinas** [comment]
Hi @ehartford , have you managed to find solution for converting fine-tuned Mistral-Nemo to 16bit gguf? I encountered the same problem you described.

**Comment by ehartford** [comment]
Yes I got past it by adding my model to the -update.py and running it then passing --model-name

**Comment by AllyourBaseBelongToUs** [comment]
> Yes I got past it by adding my model to the -update.py and running it then passing --model-name

Hi Eric, love your Dolphin models.

Are the Hugging Face GGUF models not updated?

Causing the error: pre-tokenizer type ‘dolphin12b not recognized.

The model files also show the dolphin12b tokenizer instead of Tekken :(

Any way we can help you there?

P.S.

Your models work better with your systemprompts instead of jailbreaks <3

**Comment by ehartford** [comment]
it works for me, and on ollama too
https://ollama.com/CognitiveComputations/dolphin-mistral-nemo

**Comment by ehartford** [comment]
Yes definitely - I'm totally overwhelmed

**Comment by rmusser01** [comment]
> Yes definitely - I'm totally overwhelmed

Would you mind elaborating on what might be helpful for you/how people can help?

**Comment by netrunnereve** [comment]
Does anyone know if the FP8 QAT used by Nemo is in E4M3 or E5M2? My guess is E4M3 but I couldn't find info on that anywhere, with Mistral only saying that they used FP8.

**Comment by Djip007** [comment]
I can't find what Mistral use and how (can be nice to know!!!)
The only thing I can find is:
https://huggingface.co/neuralmagic/Mistral-Nemo-Instruct-2407-FP8/tree/main?show_file_info=model-00001-of-00003.safetensors
That use E4M3 + weight_scale&input_scale

**Comment by netrunnereve** [comment]
That seems to be a vLLM/Neural Magic quant format, which they also [use for Llama](https://huggingface.co/neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8/tree/main?show_file_info=model.safetensors.index.json) as well. It's a generic quantization algorithm like our Q8_0 and isn't necessarily the format Mistral trained with.

**Comment by HyunjoonCho** [comment]
I am not sure whether this would be the proper thread to ask this - truly sorry if not,
is there any future plan to support Mistral-NeMo-Base-2407 officially?

I found some user uploaded versions but could not find one from the [official page](https://ollama.com/library/mistral-nemo/tags).

Thanks!

**Comment by EliEron** [comment]
> I am not sure whether this would be the proper thread to ask this - truly sorry if not, is there any future plan to support Mistral-NeMo-Base-2407 officially?
> 
> I found some user uploaded versions but could not find one from the [official page](https://ollama.com/library/mistral-nemo/tags).
> 
> Thanks!

Ollama is an entirely seperate project from llama.cpp. While they use llama.cpp for inference there is no official partnership between the projects so this is indeed the wrong place to ask.

I'd suggest posting in the Ollama repo instead if you feel strongly about it, but i suspect you won't get a lot of traction. Base models aren't usually considered a high priority, and most just use third party uploads.

**Comment by Djip007** [comment]
> Does anyone know if the FP8 QAT used by Nemo is in E4M3 or E5M2? My guess is E4M3 but I couldn't find info on that anywhere, with Mistral only saying that they used FP8.

I still don't know what they used. But, I'm starting to see what you can get with the different formats. [fp8](https://github.com/Mozilla-Ocho/llamafile/discussions/549#discussioncomment-10988233)

And for now the "best" is with E3M4 (yes I know it is not GPU native :sunglasses: )

**Comment by Djip007** [comment]
OK I have a POC for base FP8 support: https://github.com/ggerganov/llama.cpp/pull/10055

**Comment by netrunnereve** [comment]
@Djip007 Very nice. I actually played around with FP8 last month by imagining that the Q8_0 int8 values were actually FP8 and writing ggml_vec_dot routines for it. It ended up being much slower than Q8_0 as I had to:

1. Turn everything into F32 or F16 first using either bit fiddling or a LUT
2. Do the math as F32 (which sucks compared to say Q8_0, where the dot product can be computed much more efficiently using an 8-bit madd).

I'm not sure how fast the code in your PR runs with OMP SIMD but I'm reading through your llamafile discussion https://github.com/Mozilla-Ocho/llamafile/discussions/549. Fully optimized and with AVX512 it looks like you're getting Q8_0 performance with FP8.

**Comment by Djip007** [comment]
> Fully optimized and with AVX512 it looks like you're getting Q8_0 performance with FP8.

I am faster on prompt-processing that Q8_0 :wink: (on a zen4 CPU 42pp vs 67pp)
But to be fair we can go faster with the Q8 (on CPU). But we need more "complicated" matmul. And possible new backend.
hop to see a avx512_fp8_dot4 on zen6 ;)

**Comment by netrunnereve** [comment]
Well to be fair you've got hardware BF16 support and the dot product instructions which I think help a lot. For older computers with only AVX or AVX2 support there's a huge advantage to Q8_0 since you can work on the 8 bit weights using integer operations.

As for FP8 support yeah that's probably coming to CPU eventually, hence why I support getting this ready in llama.cpp. The next step is FP4 and Nvidia's already implementing that in their latest GPUs.

**Comment by mitar** [comment]
Still relevant though.
